calyx/opt/src/passes/simplify_invoke_with.rs

```diff
@@ -0,0 +1,163 @@
+        _sigs: &LibrarySignatures,
+        _comps: &[calyx_ir::Component],
+    ) -> VisResult {
+        if let Some(cg) = &s.comb_group {
+            let cg_name = cg.borrow().name();
+            let cell_name = s.comp.borrow().name();
+            let entry = self.comb_groups_to_modify.entry(cg_name);
+            match entry {
+                std::collections::hash_map::Entry::Occupied(occupied_entry) => {
+                    // if the two invokes involve the same cell and comb group, then ignore
+                    if !(*occupied_entry.key() == cg_name
+                        && *occupied_entry.get() == cell_name)
+                    {
+                        // there is a different invoke that is using the same comb group
+                        self.comb_groups_used_elsewhere.insert(cg_name);
+                    }
+                }
+                std::collections::hash_map::Entry::Vacant(_) => {
+                    // no invokes have used this comb group so far
+                    self.comb_groups_to_modify.insert(cg_name, cell_name);
+                }
+            }
+        }
+        Ok(Action::Continue)
+    }
+
+    fn start_while(
+        &mut self,
+        s: &mut calyx_ir::While,
+        _comp: &mut calyx_ir::Component,
+        _sigs: &LibrarySignatures,
+        _comps: &[calyx_ir::Component],
+    ) -> VisResult {
+        if let Some(comb_group) = &s.cond {
+            self.comb_groups_used_elsewhere
+                .insert(comb_group.borrow().name());
+        }
+        Ok(Action::Continue)
+    }
+
+    fn start_if(
+        &mut self,
+        s: &mut calyx_ir::If,
+        _comp: &mut calyx_ir::Component,
+        _sigs: &LibrarySignatures,
+        _comps: &[calyx_ir::Component],
+    ) -> VisResult {
+        if let Some(comb_group) = &s.cond {
+            self.comb_groups_used_elsewhere
+                .insert(comb_group.borrow().name());
+        }
+        Ok(Action::Continue)
+    }
+
+    fn finish(
+        &mut self,
+        comp: &mut calyx_ir::Component,
+        sigs: &LibrarySignatures,
+        _comps: &[calyx_ir::Component],
+    ) -> VisResult {
+        let mut builder = ir::Builder::new(comp, sigs);
+        let zero = builder.add_constant(0, 1);
+        // first, drop any comb groups that are used in while/ifs and in multiple invokes
+        for used_comb_group in &self.comb_groups_used_elsewhere {
+            self.comb_groups_to_modify.remove(used_comb_group);
+        }
+        // modify assignments of any remaining comb groups
+        for comb_group_ref in comp.comb_groups.iter() {
+            let mut comb_group = comb_group_ref.borrow_mut();
+            let comb_group_name = comb_group.name();
+            if self.comb_groups_to_modify.contains_key(&comb_group_name) {
+                let cell_name =
+                    self.comb_groups_to_modify.get(&comb_group_name).unwrap();
+                let mut modified_asgns = comb_group.assignments.clone();
```

Comment:
Recommend use of `std::mem::take` here over cloning since you're going to
replace the assignments anyway